src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt

```diff
@@ -44,50 +44,79 @@ class VisitService(
 ) {
 
   fun reserveVisitSlot(useReference: String = "", reserveVisitSlotDto: ReserveVisitSlotDto): VisitDto {
-    val visitEntity = visitRepository.saveAndFlush(
-      Visit(
-        prisonerId = reserveVisitSlotDto.prisonerId,
-        prisonId = reserveVisitSlotDto.prisonId,
-        visitRoom = reserveVisitSlotDto.visitRoom,
-        visitType = reserveVisitSlotDto.visitType,
-        visitStatus = RESERVED,
-        visitRestriction = reserveVisitSlotDto.visitRestriction,
-        visitStart = reserveVisitSlotDto.startTimestamp,
-        visitEnd = reserveVisitSlotDto.endTimestamp,
-        _reference = useReference
+
+    val bookedVisit = this.visitRepository.findBookedVisit(useReference)
+
+    if (isVisitNew(bookedVisit, reserveVisitSlotDto)) {
+
+      val visitEntity = visitRepository.saveAndFlush(
+        Visit(
+          prisonerId = reserveVisitSlotDto.prisonerId,
+          prisonId = reserveVisitSlotDto.prisonId,
+          visitRoom = reserveVisitSlotDto.visitRoom,
+          visitType = reserveVisitSlotDto.visitType,
+          visitStatus = RESERVED,
+          visitRestriction = reserveVisitSlotDto.visitRestriction,
+          visitStart = reserveVisitSlotDto.startTimestamp,
+          visitEnd = reserveVisitSlotDto.endTimestamp,
+          _reference = useReference
+        )
       )
-    )
 
-    reserveVisitSlotDto.visitContact?.let {
-      visitEntity.visitContact = createVisitContact(visitEntity, it.name, it.telephone)
-    }
+      reserveVisitSlotDto.visitContact?.let {
+        visitEntity.visitContact = createVisitContact(visitEntity, it.name, it.telephone)
+      }
 
-    reserveVisitSlotDto.visitors.forEach {
-      visitEntity.visitors.add(createVisitVisitor(visitEntity, it.nomisPersonId, it.visitContact))
-    }
+      reserveVisitSlotDto.visitors.forEach {
+        visitEntity.visitors.add(createVisitVisitor(visitEntity, it.nomisPersonId, it.visitContact))
+      }
 
-    reserveVisitSlotDto.visitorSupport?.let { supportList ->
-      supportList.forEach {
-        if (!supportTypeRepository.existsByName(it.type)) {
-          throw SupportNotFoundException("Invalid support ${it.type} not found")
+      reserveVisitSlotDto.visitorSupport?.let { supportList ->
+        supportList.forEach {
+          if (!supportTypeRepository.existsByName(it.type)) {
+            throw SupportNotFoundException("Invalid support ${it.type} not found")
+          }
+          visitEntity.support.add(createVisitSupport(visitEntity, it.type, it.text))
         }
-        visitEntity.support.add(createVisitSupport(visitEntity, it.type, it.text))
       }
+
+      createVisitTrackEvent(visitEntity)
+      return VisitDto(visitEntity)
+    } else {
+      return changeVisitSlot(bookedVisit!!.applicationReference, mapToChangeVisitSlotRequestDto(reserveVisitSlotDto))
```

Comment:
Presuming bookedVisit will be not null as we have done the check in if clause. Is that right?

Reply:
It's never null at this stage